library/src/main/java/com/google/android/exoplayer2/extractor/mp4/AtomParsers.java

```diff
@@ -423,47 +430,333 @@ private static void parseMetaAtom(ParsableByteArray data, GaplessInfoHolder out)
+      ilst.skipBytes(4);
+      String value = ilst.readNullTerminatedString(length - 4);
+      Id3Frame frame = new TextInformationFrame(attributeName, value);
+      builder.add(frame);
+    } else {
+      ilst.skipBytes(length);
+    }
+  }
+
+  private static void parseCommentAttribute(MetadataBuilder builder,
+                                            String attributeName,
+                                            ParsableByteArray ilst,
+                                            int endPosition) {
+    int length = ilst.readInt() - Atom.FULL_HEADER_SIZE;
+    int key = ilst.readInt();
+    ilst.skipBytes(4);
+    if (key == Atom.TYPE_data) {
+      ilst.skipBytes(4);
+      String value = ilst.readNullTerminatedString(length - 4);
+      Id3Frame frame = new CommentFrame("eng", attributeName, value);
+      builder.add(frame);
+    } else {
+      ilst.skipBytes(length);
+    }
+  }
+
+  private static void parseBooleanAttribute(MetadataBuilder builder,
+                                            String attributeName,
+                                            ParsableByteArray ilst,
+                                            int endPosition) {
+    int length = ilst.readInt() - Atom.FULL_HEADER_SIZE;
+    int key = ilst.readInt();
+    ilst.skipBytes(4);
+    if (key == Atom.TYPE_data) {
+      Object value = parseDataBox(ilst, length);
+      if (value instanceof Integer) {
+        int n = (Integer) value;
+        String s = n == 0 ? "0" : "1";
+        Id3Frame frame = new TextInformationFrame(attributeName, s);
+        builder.add(frame);
+      }
+    } else {
+      ilst.skipBytes(length);
+    }
+  }
+
+  private static void parseIntegerAttribute(MetadataBuilder builder,
+                                            String attributeName,
+                                            ParsableByteArray ilst,
+                                            int endPosition) {
+    int length = ilst.readInt() - Atom.FULL_HEADER_SIZE;
+    int key = ilst.readInt();
+    ilst.skipBytes(4);
+    if (key == Atom.TYPE_data) {
+      Object value = parseDataBox(ilst, length);
+      if (value instanceof Integer) {
+        int n = (Integer) value;
+        String s = "" + n;
+        Id3Frame frame = new TextInformationFrame(attributeName, s);
+        builder.add(frame);
+      }
+    } else {
+      ilst.skipBytes(length);
+    }
+  }
+
+  private static void parseIndexAndCountAttribute(MetadataBuilder builder,
+                                                  String attributeName,
+                                                  ParsableByteArray ilst,
+                                                  int endPosition) {
+    int length = ilst.readInt() - Atom.FULL_HEADER_SIZE;
+    int key = ilst.readInt();
+    ilst.skipBytes(4);
+    if (key == Atom.TYPE_data) {
+      Object value = parseDataBox(ilst, length);
+      if (value instanceof byte[]) {
+        byte[] bytes = (byte[]) value;
+        if (bytes.length == 8) {
```

Comment:
I have a piece of sample media where the length is only 6. I couldn't find an authoritative source to say whether this is valid, but since the final two bytes are ignored regardless, it seemed fine to handle this case.